Spec/PushActivationStateMachine.swift

```diff
@@ -654,77 +660,772 @@ class PushActivationStateMachine : QuickSpec {
+                var storage: MockDeviceStorage!
+
+                beforeEach {
+                    storage = MockDeviceStorage(startWith: ARTPushActivationStateWaitingForDeregistration(machine: initialStateMachine))
+                    rest.storage = storage
+                    stateMachine = ARTPushActivationStateMachine(rest)
+                }
+
+                // RSH3g1
+                it("on Event CalledDeactivate") {
+                    stateMachine.send(ARTPushActivationEventCalledDeactivate())
+                    expect(stateMachine.current).to(beAKindOf(ARTPushActivationStateWaitingForDeregistration.self))
+                }
+
+                // RSH3g2
+                it("on Event Deregistered") {
+                    var deactivatedCallbackCalled = false
+                    let hook = stateMachine.testSuite_injectIntoMethod(after: NSSelectorFromString("callDeactivatedCallback:")) {
+                        deactivatedCallbackCalled = true
+                    }
+                    defer { hook.remove() }
+
+                    var setAndPersistIdentityTokenDetailsCalled = false
+                    let hookDevice = stateMachine.rest.device.testSuite_injectIntoMethod(after: NSSelectorFromString("setAndPersistIdentityTokenDetails:")) {
+                        setAndPersistIdentityTokenDetailsCalled = true
+                    }
+                    defer { hookDevice.remove() }
+
+                    stateMachine.send(ARTPushActivationEventDeregistered())
+                    expect(stateMachine.current).to(beAKindOf(ARTPushActivationStateNotActivated.self))
+                    expect(deactivatedCallbackCalled).to(beTrue())
+                    expect(setAndPersistIdentityTokenDetailsCalled).to(beTrue())
+                    // RSH3g2a
+                    expect(stateMachine.rest.device.identityTokenDetails).to(beNil())
+                }
+
+                // RSH3g3
+                it("on Event DeregistrationFailed") {
+                    let expectedError = ARTErrorInfo(domain: ARTAblyErrorDomain, code: 1234, userInfo: nil)
+
+                    var deactivatedCallbackCalled = false
+                    let hook = stateMachine.testSuite_getArgument(from: NSSelectorFromString("callDeactivatedCallback:"), at: 0, callback: { arg0 in
+                        deactivatedCallbackCalled = true
+                        guard let error = arg0 as? ARTErrorInfo else {
+                            fail("Error is missing"); return
+                        }
+                        expect(error) == expectedError
+                    })
+                    defer { hook.remove() }
+
+                    stateMachine.send(ARTPushActivationEventDeregistrationFailed(error: expectedError))
+                    expect(stateMachine.current).to(beAKindOf(ARTPushActivationStateWaitingForDeregistration.self))
+                    expect(deactivatedCallbackCalled).to(beTrue())
+                }
```

Comment:
I think this is missing a test for [RSH3g3b](https://docs.ably.io/client-lib-development-guide/features/#RSH3g3b)